### PR TITLE
iOS: In case Xcode is installed but the version is too old. Advise the user to update.

### PR DESCRIPTION
**Pull request by chinmaygarde**
- The tools will refuse to function with an old version of Xcode.
- doctor will diagnose if the Xcode version is satisfactory.

### CODE REVIEW

**Comment by chinmaygarde** [comment]
Fixes https://github.com/flutter/flutter/issues/1985

**Comment by abarth** [comment]
@devoncarew

File: packages/flutter_tools/lib/src/ios/ios_workflow.dart [hunk 1]
```diff
@@ -44,11 +61,20 @@ class IOSWorkflow extends Workflow {
...
       validatorFunction: _xcodeExists
+    );
+
+    iosValidator.addValidator(xcodeValidator);
+
+    xcodeValidator.addValidator(new Validator(
+      'Version Check',
+      description: 'Xcode version is at least $kXcodeRequiredVersionMajor.$kXcodeRequiredVersionMinor',
+      resolution: 'Download the latest version or update via the Mac App Store',
+      validatorFunction: _xcodeVersionSatisfactory
```
Comment (abarth) [inline packages/flutter_tools/lib/src/ios/ios_workflow.dart]: You can do this slightly fancier with the `..` operator.

File: packages/flutter_tools/lib/src/ios/mac.dart [hunk 2]
```diff
@@ -5,10 +5,49 @@
 import '../base/context.dart';
 import '../base/process.dart';
 
+const int kXcodeRequiredVersionMajor = 7;
+const int kXcodeRequiredVersionMinor = 2;
```
Comment (abarth) [inline packages/flutter_tools/lib/src/ios/mac.dart]: Should we make these private by naming them with a leading `_` ?

**Comment by abarth** [comment]
LGTM

File: packages/flutter_tools/lib/src/ios/ios_workflow.dart [hunk 3]
```diff
@@ -27,7 +28,23 @@ class IOSWorkflow extends Workflow {
     );
 
     Function _xcodeExists = () {
-      return xcode.isInstalled ? ValidationType.installed : ValidationType.missing;
+      if (xcode.isInstalledAndMeetsVersionCheck) {
```
Comment (devoncarew) [inline packages/flutter_tools/lib/src/ios/ios_workflow.dart]: here and elsewhere, one line `if`s should omit curly braces

File: packages/flutter_tools/lib/src/ios/mac.dart [hunk 4]
```diff
@@ -5,10 +5,49 @@
...
+      int minor = components.length == 1 ? 0 : int.parse(components[1]);
+
+      _xcodeVersionSatisfactory = major >= kXcodeRequiredVersionMajor && minor >= kXcodeRequiredVersionMinor;
+      return _xcodeVersionSatisfactory;
+    } catch (error) {
+      _xcodeVersionSatisfactory = false;
+      return false;
+    }
+
+    return false;
```
Comment (devoncarew) [inline packages/flutter_tools/lib/src/ios/mac.dart]: It looks like this return is dead code. When doing these type of calc-once caching, I tend to only return at the end:

```
if (_value == null) {
  _value = ... // calc it
}

return _value;
```

Comment (chinmaygarde) [inline packages/flutter_tools/lib/src/ios/mac.dart]: ugh, the version check is bogus. Will fix both. Thanks for making me look again :)

**Pull request merged**